Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -225,7 +296,29 @@ public struct HTTPHeaders: CustomStringConvertible {
     ///
     /// - Parameter name: The name of the header field to remove from the block.
     public mutating func remove(name: String) {
-        self.storage[name.lowercased()] = nil
+        guard !self.indices.isEmpty else {
+            return
+        }
+
+        let utf8 = name.utf8
+        let stringLength = utf8.count
+        var array: [Int] = []
+        // We scan from the back to the front so we can remove the subranges with as less overhead as possible.
+        for idx in stride(from: self.indices.count - 2, to: -1, by: -2) {
+            let nameIdx = self.indices[idx]
+            if stringLength == nameIdx.length && self.buffer.equalCaseInsensitiveASCII(seq: utf8, at: nameIdx) {
+                array.append(idx)
+            }
+        }
+
+        guard !array.isEmpty else {
+            return
+        }
+        array.forEach {
+            self.indices.removeSubrange($0...$0 + 1)
```

Comment:
`removeSubrange`:
> Complexity: O(n), where n is the length of the collection.

I used an array of optional header indexes in my implementation so that removed values could just be set to `nil` without having to re-arrage the array.

Reply:
@tanner0101 in my first iteration I did exactly this but it had some performance impact (most likely because of all the extra branching that was needed while iterating though the headers.

With the current code:
```
✗ wrk -c 256 -d 10s -t 4 -s ~/Downloads/pipeline-many.lua -H "X-Host: SomeValue" -H "Host: swiftnio.io" -H "ThereAreEvenMoreHeaders: AndMoreValues" http://localhost:8888
Running 10s test @ http://localhost:8888
  4 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   124.51ms  182.45ms   1.99s    89.44%
    Req/Sec    78.90k    27.18k  210.27k    82.01%
  3065753 requests in 10.05s, 146.19MB read
  Socket errors: connect 0, read 179, write 0, timeout 104
Requests/sec: 304981.98
Transfer/sec:     14.54MB
```

With using optionals:
```
✗ wrk -c 256 -d 10s -t 4 -s ~/Downloads/pipeline-many.lua -H "X-Host: SomeValue" -H "Host: swiftnio.io" -H "ThereAreEvenMoreHeaders: AndMoreValues" http://localhost:8888
Running 10s test @ http://localhost:8888
  4 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   117.40ms  149.18ms   2.00s    90.74%
    Req/Sec    71.97k    23.04k  165.79k    71.54%
  2854242 requests in 10.06s, 136.10MB read
  Socket errors: connect 0, read 160, write 0, timeout 122
Requests/sec: 283653.24
Transfer/sec:     13.53MB
```

As you know its all about trade-offs but the question is how often people will remove headers vs   trying to access some. Also note its worst case 0(n) as often it should be able to do much better as the sub-range you remove may be at the end of the array.

These numbers can be reproduced all the time.

Also just for the record this is the performance of current master (without this pr at all):

```
 ✗ wrk -c 256 -d 10s -t 4 -s ~/Downloads/pipeline-many.lua -H "X-Host: SomeValue" -H "Host: swiftnio.io" -H "ThereAreEvenMoreHeaders: AndMoreValues" http://localhost:8888
Running 10s test @ http://localhost:8888
  4 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   138.82ms  147.82ms   1.99s    85.21%
    Req/Sec    51.70k    21.46k  109.23k    65.57%
  2042977 requests in 10.09s, 97.42MB read
  Socket errors: connect 0, read 49, write 9, timeout 136
Requests/sec: 202409.96
Transfer/sec:      9.65MB
```

Reply:
Also note with the latest commit (f533f4d193a3db6f5c958db6ba9d638bb2dcf887) we are now doing:

```
wrk -c 256 -d 10s -t 4 -s ~/Downloads/pipeline-many.lua -H "X-Host: SomeValue" -H "Host: swiftnio.io" -H "ThereAreEvenMoreHeaders: AndMoreValues" http://localhost:8888
Running 10s test @ http://localhost:8888
  4 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   110.49ms  148.06ms   1.95s    93.29%
    Req/Sec    81.94k    24.12k  201.77k    76.19%
  3265185 requests in 10.05s, 155.70MB read
  Socket errors: connect 0, read 112, write 0, timeout 114
Requests/sec: 325043.36
Transfer/sec:     15.50MB
```

Less array access FTW ;)

Reply:
oh wow, that's awesome! 👍